### PR TITLE
Auto assign GPUs for NIM deployments when a cached optimized profile is specified

### PR DESCRIPTION
**Pull request by shivamerla**
Optimized NIM profile config contains a field called tensorParallelism which indicates required number of GPUs. Use that to auto-assign GPU resources for NIM deployments.

### CODE REVIEW

File: internal/controller/platform/standalone/nimservice.go [hunk 1]
```diff
@@ -413,3 +427,82 @@ func (r *NIMServiceReconciler) reconcilePVC(ctx context.Context, nimService *app
...
+			}
+		}
+
+		// TODO: Make the resource name configurable
+		const gpuResourceName = corev1.ResourceName("nvidia.com/gpu")
+
+		deploymentParams.Resources.Requests[corev1.ResourceName(gpuResourceName)] = gpuQuantity
+		deploymentParams.Resources.Limits[corev1.ResourceName(gpuResourceName)] = gpuQuantity
+	}
+
```
Comment (slu2011) [inline internal/controller/platform/standalone/nimservice.go:506]: How about add some logging when tensorParallelism is empty?

File: internal/controller/platform/standalone/nimservice.go [hunk 2]
```diff
@@ -413,3 +427,82 @@ func (r *NIMServiceReconciler) reconcilePVC(ctx context.Context, nimService *app
...
+		logger.Error(err, "Failed to retrieve tensorParallelism")
+		return err
+	}
+
+	// Check if tensorParallelism is defined in the profile, and automatically assign GPU resources.
+	// Note: This will override any manual GPU assignments made by the user.
+	// The number of GPUs for an optimized profile is fixed.
+	// Allocating more GPUs than required may result in underutilization,
+	// while allocating fewer GPUs will likely cause failures.
+	if tensorParallelism != "" {
```
Comment (slu2011) [inline internal/controller/platform/standalone/nimservice.go:486]: if tensorParallelism is empty, then the user specified gpu resource count would be used? Or should it default to 1 gpu?

File: internal/controller/platform/standalone/nimservice.go (hunk 1 above)
Comment (shivamerla) [reply internal/controller/platform/standalone/nimservice.go:534]: done.

File: internal/controller/platform/standalone/nimservice.go (hunk 2 above)
Comment (shivamerla) [reply internal/controller/platform/standalone/nimservice.go:524]: Good point. I have updated the logic a bit to honor the request from the user. If none are specified, then auto-assignment is done. This is all when a NIM profile is selected.

File: internal/controller/platform/standalone/nimservice.go [hunk 3]
```diff
@@ -459,50 +459,82 @@ func (r *NIMServiceReconciler) getNIMCacheProfile(ctx context.Context, nimServic
 	return nil, nil
 }
 
+// getTensorParallelismByProfile returns the value of tensor parallelism parameter in the given NIM profile
 func (r *NIMServiceReconciler) getTensorParallelismByProfile(ctx context.Context, profile *appsv1alpha1.NIMProfile) (string, error) {
+	// List of possible keys for tensor parallelism
+	possibleKeys := []string{"tensorParallelism", "tp"}
```
Comment (slu2011) [inline internal/controller/platform/standalone/nimservice.go:465]: :)

**Review by slu2011 — APPROVED** [review]

File: internal/controller/platform/standalone/nimservice.go (hunk 3 above)
Comment (shivamerla) [reply internal/controller/platform/standalone/nimservice.go:465]: we never know :)

**Review by visheshtanksale — APPROVED** [review]
LGTM

**Pull request merged**